riscv/src/register/macros.rs

```diff
@@ -600,3 +600,979 @@ macro_rules! clear_pmp {
+
+                     if $field_end < max_width && $field_start < $field_end {
+                         let value = $crate::bits::bf_extract(
+                             self.bits,
+                             $field_start,
+                             $field_end - $field_start + 1,
+                         );
+
+                         $field_ty::from_usize(value).ok_or($crate::result::Error::InvalidFieldVariant {
+                             field: stringify!($field),
+                             value,
+                         })
+                     } else {
+                        Err($crate::result::Error::IndexOutOfBounds {
+                            index: $field_start,
+                            min: $field_start,
+                            max: $field_end,
+                        })
+                     }
+                 }
+             }
+         }
+    };
+}
+
+/// Defines field accesor functions for a write-only CSR type.
+#[macro_export]
+macro_rules! write_only_csr_field {
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     bit: $bit:literal$(,)?) => {
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&mut self, $field: bool) {
+                self.$try_field($field).unwrap();
+            }
+
+            $(#[$try_field_doc])+
+            #[inline]
+            pub fn $try_field(&mut self, $field: bool) -> $crate::result::Result<()> {
+                let max_width = core::mem::size_of_val(&self.bits) * 8;
+
+                if $bit < max_width {
+                    self.bits = $crate::bits::bf_insert(self.bits, $bit, 1, $field as usize);
+                    Ok(())
+                } else {
+                    Err($crate::result::Error::IndexOutOfBounds {
+                        index: $bit,
+                        min: 0,
+                        max: max_width,
+                    })
+                }
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     range: $bit_start:literal..=$bit_end:literal$(,)?) => {
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&mut self, index: usize, $field: bool) {
+                self.$try_field(index, $field).unwrap();
+            }
+
+            $(#[$try_field_doc])+
+            #[inline]
+            pub fn $try_field(&mut self, index: usize, $field: bool) -> $crate::result::Result<()> {
+                let max_width = core::mem::size_of_val(&self.bits) * 8;
+
+                if index < max_width && ($bit_start..=$bit_end).contains(&index) {
+                    self.bits = $crate::bits::bf_insert(self.bits, index, 1, $field as usize);
+                    Ok(())
+                } else {
+                    Err($crate::result::Error::IndexOutOfBounds {
```

Comment:
Here more or less the same. We need this `try_field` method, as we don't know if the index will be valid. But we can be sure at compile time that `$bit_start` and `$bit_end` are correct, so we just need to check that `index` is inside this range.

Reply:
Similarly as above, can you provide an example of what the compile-time check would look like?

Reply:
Same as above. We add compile-time checks for `start` and `end`. Then, we only need to check that the parameter index is within the expected range.